packages/node/src/integrations/context.ts

```diff
@@ -0,0 +1,334 @@
+/* eslint-disable max-lines */
+import {
+  AppContext,
+  Contexts,
+  CultureContext,
+  DeviceContext,
+  Event,
+  EventProcessor,
+  Integration,
+  OsContext,
+} from '@sentry/types';
+import { execFile } from 'child_process';
+import { readdir, readFile } from 'fs';
+import * as os from 'os';
+import { join } from 'path';
+import { promisify } from 'util';
+
+// TODO: Required until we drop support for Node v8
+export const readFileAsync = promisify(readFile);
+export const readDirAsync = promisify(readdir);
+
+interface DeviceContextOptions {
+  cpu?: boolean;
+  memory?: boolean;
+}
+
+interface ContextOptions {
+  app?: boolean;
+  os?: boolean;
+  device?: DeviceContextOptions | boolean;
+  culture?: boolean;
+}
+
+/** Add node modules / packages to the event */
+export class Context implements Integration {
+  /**
+   * @inheritDoc
+   */
+  public static id: string = 'Context';
+
+  /**
+   * @inheritDoc
+   */
+  public name: string = Context.id;
+
+  /**
+   * Caches OS context so it's only evaluated once
+   */
+  private _cachedOsContext: Promise<OsContext> | undefined;
+
+  public constructor(private readonly _options: ContextOptions = { app: true, os: true, device: true, culture: true }) {
+    //
```

Comment:
What do you think about generating the context in the constructor

```
public constructor(private readonly _options: ContextOptions = { app: true, os: true, device: true, culture: true }) {
  // can make `addContext` sync because it's only run once
  this._cachedContext = this.getContext()
}
```

and then change `setupOnce` to be:

```ts
addGlobalEventProcessor(event => { ...event, contexts: { ...event.contexts, ..._cachedContexts } });
```

Reply:
We can't cache all the context because some of it will change at runtime. ie. `free_memory`, `memory_used` and possibly even `processor_frequency` depending on whether Node.js gets the processor max or current freq.

For now I have just cached the `os` because in some cases that involves reading files, etc. 

We could cache everything apart from those dynamic properties?

Reply:
Yeah that sounds fair to me!